services/subscription.go

```diff
@@ -0,0 +1,187 @@
+package services
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/smartcontractkit/chainlink/logger"
+	"github.com/smartcontractkit/chainlink/store"
+	"github.com/smartcontractkit/chainlink/store/models"
+	"github.com/smartcontractkit/chainlink/store/presenters"
+	"github.com/smartcontractkit/chainlink/utils"
+)
+
+const (
+	EventTopicSignature = iota
+	EventTopicRequestID
+	EventTopicJobID
+)
+
+// RunLogTopic is the signature for the Request(uint256,bytes32,string) event
+// which Chainlink RunLog initiators watch for.
+// See https://github.com/smartcontractkit/chainlink/blob/master/solidity/contracts/Oracle.sol
+var RunLogTopic = common.HexToHash("0x06f4bf36b4e011a5c499cef1113c2d166800ce4013f6c2509cab1a0e92b83fb2")
+
+// Listens to event logs being pushed from the Ethereum Node specific to this job.
+type Subscription struct {
+	Job              models.Job
+	store            *store.Store
+	logNotifications chan types.Log
+	errors           chan error
+	rpcSubscription  *rpc.ClientSubscription
+}
+
+// Constructor of Subscription that to starts listening to and keeps track of
+// event logs corresponding to a job.
+func StartSubscription(job models.Job, store *store.Store) (Subscription, error) {
+	var addresses []common.Address
+	for _, initr := range job.InitiatorsFor(models.InitiatorEthLog, models.InitiatorRunLog) {
```

Comment:
Hi @se3000, imagine the follow job json:

```json
{
  "initiators": [
    { "type": "ethlog", "address": "0x3cCad4715152693fE3BC4460591e3D3Fbd071b42" },
    { "type": "runlog" }
  ],
  ...
}
```
Note that the `runlog` will default to and address of 0x0000 that we cull out.

The issue here is that when creating our ws subscription, we'll only create it for the `ethlog` address and ignore listening to all addresses for `runlog`. Given how `eth_newFilter` works (https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_newfilter), this isn't possible with one subscription. I don't think we should have to manage multiple subscriptions per job for the time being so I'm leaning towards only allowing one log initiator per job.

Thoughts?

Reply:
Discussed offline. Solution would be a subscription per initiator but it is indeed an edge case.